### ISSUE TITLE
How to check the total no of neurons of model?

### ISSUE THREAD

**Issue by neso613** [tensorflow/tensorflow#53644]
My question is about the Neural Network neuron's numbers.
No of neurons in input layer is equal to no of input features..say 32,32,1 if MNIST dataset 
and no of neurons in Output layer is equal to no of target variable say,,in MNIST it is 10
but how can we know about the neurons of hidden layers?
https://colab.research.google.com/github/AviatorMoser/keras-mnist-tutorial/blob/master/MNIST%20in%20Keras.ipynb#scrollTo=BL0AhUGJwT6M
in this colab notebook, there are 2 network defined.
How can we know the total no of neurons of the architecure?

Same question is for TFLite model too.
Also, How to check the performance in terms of TOPs/Power consumption/GPU-CPU usage etc?

**Comment by mohantym** [comment]
Hi @neso613 ! Can you check these  threads [1](https://stackoverflow.com/questions/56518924/how-we-can-get-the-total-number-of-neurons-from-a-model-in-keras-is-there-any-f) ,[2 ](https://stackoverflow.com/questions/54408347/count-neurons-in-keras-with-different-layers-is-my-approach-correct) for counting neurons in a keras model ? You can check memory usage of CPU/GPU  using [memory_usage](https://www.tensorflow.org/api_docs/python/tf/config/experimental/get_memory_usage) api . You can use a USB powermeter along your micro-controllers to check the power usage. Please post in TF forum/SO for further assistance. Thank you!

**Comment by Anindyadeep** [comment]
I will assume that, you already have some knowledge about Neural Nets. So, when you write this:
```
model.add(Dense(512, input_shape=(784,)))
```
This line means your Input neurons is = total number of pixels in MNIST (here)
And this is `connected` to a hidden layer of `512` neurons. So the next hidden layer `h1` will have `512` neurons. Now suppose after this you add this line.
```
model.add(Dense(64, activation ="relu")
```
So, this means that the next layer `h2` will have `64` number of neurons. And `h1` is connected to `h2`. And this is how it goes on.
So, the first hidden layer has `h1` has `512` neurons. Second layer `h2` has `64` neurons. And similarly, you can count number of neurons for other layers. 

# 
About total number of neurons:

Suppose your NN architecture is defined as:
```
model.add(Dense(512, input_shape=(784,)))
model.add(Dense(64, activation ="relu")
model.add(Dense(32, activation ="relu")
model.add(Dense(10, activation ="softmax")
```
So, here Input layer `I` has `784` Neurons. Hidden Layer `h1` has `512` neurons. Hidden Layer `h2` has `64` neurons. Hidden Layer `h3` has `32` neurons. Last layer `h4` has `10` neurons. So  total number of hidden neurons in this architectures is :
```
512 + 64 + 32  = 608 neurons.
```

Now I am excluding the input and the last layer. As these two layer `input` and `last layer` is independent of our choice, as `784` neurons is due to flatten size of image, and there are `10` classes. So hidden neuron number would be here 608. There are some sources, where we see last layer (and sometime input is also considered). Hope this helps. Please feel free to correct me if I missed out at some thing.

**Comment by neso613** [comment]
> Hi @neso613 ! Can you check these threads [1](https://stackoverflow.com/questions/56518924/how-we-can-get-the-total-number-of-neurons-from-a-model-in-keras-is-there-any-f) ,[2 ](https://stackoverflow.com/questions/54408347/count-neurons-in-keras-with-different-layers-is-my-approach-correct) for counting neurons in a keras model ? You can check memory usage of CPU/GPU using [memory_usage](https://www.tensorflow.org/api_docs/python/tf/config/experimental/get_memory_usage) api . You can use a USB Power Meter along your micro-controllers to check the power usage during inference .
> 
> Please post in [TF forum](https://discuss.tensorflow.org/)/SO for further assistance. Thank you!

Thanks for reply but can USB Power Meter works with rasp?

**Comment by neso613** [comment]
> I will assume that, you already have some knowledge about Neural Nets. So, when you write this:
> 
> ```
> model.add(Dense(512, input_shape=(784,)))
> ```
> 
> This line means your Input neurons is = total number of pixels in MNIST (here) And this is `connected` to a hidden layer of `512` neurons. So the next hidden layer `h1` will have `512` neurons. Now suppose after this you add this line.
> 
> ```
> model.add(Dense(64, activation ="relu")
> ```
> 
> So, this means that the next layer `h2` will have `64` number of neurons. And `h1` is connected to `h2`. And this is how it goes on. So, the first hidden layer has `h1` has `512` neurons. Second layer `h2` has `64` neurons. And similarly, you can count number of neurons for other layers.
> 
> About total number of neurons:
> 
> Suppose your NN architecture is defined as:
> 
> ```
> model.add(Dense(512, input_shape=(784,)))
> model.add(Dense(64, activation ="relu")
> model.add(Dense(32, activation ="relu")
> model.add(Dense(10, activation ="softmax")
> ```
> 
> So, here Input layer `I` has `784` Neurons. Hidden Layer `h1` has `512` neurons. Hidden Layer `h2` has `64` neurons. Hidden Layer `h3` has `32` neurons. Last layer `h4` has `10` neurons. So total number of hidden neurons in this architectures is :
> 
> ```
> 512 + 64 + 32  = 608 neurons.
> ```
> 
> Now I am excluding the input and the last layer. As these two layer `input` and `last layer` is independent of our choice, as `784` neurons is due to flatten size of image, and there are `10` classes. So hidden neuron number would be here 608. There are some sources, where we see last layer (and sometime input is also considered). Hope this helps. Please feel free to correct me if I missed out at some thing.

So, how model parameters are different from no of neurons?

**Comment by mohantym** [comment]
@neso613 ! You can check this [thread ](https://forums.raspberrypi.com/viewtopic.php?t=288867)for Raspberry pi power meter. Thanks for the lucid explanation @Anindyadeep . You can check more on model parameters and HyperParameters from below threads. Link [1](https://www.geeksforgeeks.org/difference-between-model-parameters-vs-hyperparameters/) ,[2](https://www.kaggle.com/questions-and-answers/190624) . Thank you!

**Comment by neso613** [comment]
> @neso613 ! You can check this [thread ](https://forums.raspberrypi.com/viewtopic.php?t=288867)for Raspberry pi power meter. Thanks for the lucid explanation @Anindyadeep . You can check more on model parameters and HyperParameters from below threads. Link [1](https://www.geeksforgeeks.org/difference-between-model-parameters-vs-hyperparameters/) ,[2](https://www.kaggle.com/questions-and-answers/190624) .Thank you!

My Model parameters means, after model.summary() in Tensorflow, it shows model architecture and below to it prints abput parameter..say..trainable parameters, non-trainable parameters.

What is these trainable parameters, non-trainable parameters? Is it models neurons information?

**Comment by Anindyadeep** [comment]
Generally, what the `model.summary()` shows the architecture of the model. Like how many dense blocks u have used, how many convolutional blocks u have used etc. At the very end you will see that there is written about this much `trainable param` and this much `non-trainable` parameters. Trainable parameters are those parameters, which will be updated over time, during the time of back propagation. If you know that the hidden neurons, which we set in the dense layers, are nothing but some sort of `weight` matrix. The weights (elements of that matrix) are the trainable parameters. At the same time, suppose non-trainable params are those, which are not updated over time. Some of them are `Dropout layer`, `MaxPool layer` etc. As these parameters do not provide a learnable paradigm in the models. So this combination of trainable and non-trainable parameters, both constitutes as parameters of the model, and that whole information of that model architecture is being provided by the summary. 

Also, fun fact, Do not get confused with hyper-parameter, and non-trainable parameter. Hyper parameter are some sort of user defined values, which are tweakable and are set to the models. Like the number of hidden neurons in layers, the learning rate, the value of `p` in dropout etc. Non trainable parameters originates from the model, and have some dependency of set up of hyper parameters. Like suppose in case of dropout layer, the value of p. Or the strides taken in maxpool etc. 

Go through this [link](https://towardsdatascience.com/how-to-calculate-the-number-of-parameters-in-keras-models-710683dae0ca), to have more in-depth understanding of trainable and non-trainable parameters. 

Also, correct me, if I missed out at some thing. Thank you.
